`io.camunda.zeebe.client.impl.oauth.OAuthCredentialsCache` is not thread safe.

**Describe the bug**

<!-- A clear and concise description of what the bug is. -->
After https://github.com/camunda/zeebe/issues/11816 I never thought that the `io.camunda.zeebe.client.impl.oauth.OAuthCredentialsProvider` is not a thread-safe. So, when you start multiple `JobWorker` in parallel with `OAuthCredentialsProvider` you got a problem in this method: https://github.com/camunda/zeebe/blob/28fe054b79fcc7398a086a1576199efef01a122c/clients/java/src/main/java/io/camunda/zeebe/client/impl/oauth/OAuthCredentialsCache.java#L87-L111

So, in there is the first checking that the file exists, and then there is a creating one if none exists. If this part of the code is executed concurrently we get an error that we are trying to create a file that already exists.

**To Reproduce**

<!--
Steps to reproduce the behavior

If possible add a minimal reproducer code sample
- when using the Java client: https://github.com/zeebe-io/zeebe-test-template-java

-->

1. Get a fresh main branch.
2. Configure Zeebe Client with `OAuthCredentialsProvider`
3. Start some requests in the parallel

**Expected behavior**

<!-- A clear and concise description of what you expected to happen. -->
If requests with the Zeebe Client and `OAuthCredentialsProvider` runned at parallel there is no error about a file, that already created.  

**Log/Stacktrace**

<!-- If possible add the full stacktrace or Zeebe log which contains the issue. -->

<details><summary>Full Stacktrace</summary>
 <p>

```
io.grpc.StatusException: CANCELLED
	at io.grpc.Status.asException(Status.java:554)
	at io.grpc.kotlin.ClientCalls$rpcImpl$1$1$1.onClose(ClientCalls.kt:296)
	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:563)
	at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:70)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:744)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:723)
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:133)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	at java.base/java.lang.Thread.run(Thread.java:832)
Caused by: java.nio.file.FileAlreadyExistsException: /root/.camunda/credentials
	at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:94)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:106)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
	at java.base/sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:218)
	at java.base/java.nio.file.Files.newByteChannel(Files.java:375)
	at java.base/java.nio.file.Files.createFile(Files.java:652)
	at io.camunda.zeebe.client.impl.oauth.OAuthCredentialsCache.ensureCacheFileExists(OAuthCredentialsCache.java:110)
	at io.camunda.zeebe.client.impl.oauth.OAuthCredentialsCache.writeCache(OAuthCredentialsCache.java:69)
	at io.camunda.zeebe.client.impl.oauth.OAuthCredentialsProvider.refreshCredentials(OAuthCredentialsProvider.java:133)
	at io.camunda.zeebe.client.impl.oauth.OAuthCredentialsProvider.loadCredentials(OAuthCredentialsProvider.java:121)
	at io.camunda.zeebe.client.impl.oauth.OAuthCredentialsProvider.applyCredentials(OAuthCredentialsProvider.java:79)
```

</p>
</details>

**Environment:**
- OS: MacOS, Linux
- Zeebe Version: from main branch
- Configuration: `OAuthCredentialsProvider`

--- a/clients/java/pom.xml
+++ b/clients/java/pom.xml
@@ -84,6 +84,12 @@
     </dependency>
 
     <dependency>
+      <groupId>com.google.code.findbugs</groupId>
+      <artifactId>jsr305</artifactId>
+      <scope>provided</scope>
+    </dependency>
+
+    <dependency>
       <groupId>io.grpc</groupId>
       <artifactId>grpc-testing</artifactId>
       <scope>test</scope>

--- a/clients/java/src/main/java/io/camunda/zeebe/client/CredentialsProvider.java
+++ b/clients/java/src/main/java/io/camunda/zeebe/client/CredentialsProvider.java
@@ -20,6 +20,7 @@ import io.camunda.zeebe.client.impl.oauth.OAuthCredentialsProviderBuilder;
 import io.grpc.Metadata;
 import java.io.IOException;
 
+/** Implementations of this interface must be thread-safe. */
 public interface CredentialsProvider {
 
   /**

--- a/clients/java/src/main/java/io/camunda/zeebe/client/impl/oauth/OAuthCredentialsCache.java
+++ b/clients/java/src/main/java/io/camunda/zeebe/client/impl/oauth/OAuthCredentialsCache.java
@@ -31,7 +31,9 @@ import java.util.HashMap;
 import java.util.Map;
 import java.util.Map.Entry;
 import java.util.Optional;
+import javax.annotation.concurrent.ThreadSafe;
 
+@ThreadSafe
 public final class OAuthCredentialsCache {
 
   private static final String KEY_AUTH = "auth";
@@ -48,7 +50,7 @@ public final class OAuthCredentialsCache {
     audiences = new HashMap<>();
   }
 
-  public OAuthCredentialsCache readCache() throws IOException {
+  public synchronized OAuthCredentialsCache readCache() throws IOException {
     if (!cacheFile.exists() || cacheFile.length() == 0) {
       return this;
     }
@@ -60,7 +62,7 @@ public final class OAuthCredentialsCache {
     return this;
   }
 
-  public void writeCache() throws IOException {
+  public synchronized void writeCache() throws IOException {
     final Map<String, Map<String, OAuthCachedCredentials>> cache = new HashMap<>(audiences.size());
     for (final Entry<String, OAuthCachedCredentials> audience : audiences.entrySet()) {
       cache.put(audience.getKey(), Collections.singletonMap(KEY_AUTH, audience.getValue()));
@@ -70,17 +72,17 @@ public final class OAuthCredentialsCache {
     MAPPER.writer().writeValue(cacheFile, cache);
   }
 
-  public Optional<ZeebeClientCredentials> get(final String endpoint) {
+  public synchronized Optional<ZeebeClientCredentials> get(final String endpoint) {
     return Optional.ofNullable(audiences.get(endpoint)).map(OAuthCachedCredentials::getCredentials);
   }
 
-  public OAuthCredentialsCache put(
+  public synchronized OAuthCredentialsCache put(
       final String endpoint, final ZeebeClientCredentials credentials) {
     audiences.put(endpoint, new OAuthCachedCredentials(credentials));
     return this;
   }
 
-  public int size() {
+  public synchronized int size() {
     return audiences.size();
   }
 

--- a/clients/java/src/main/java/io/camunda/zeebe/client/impl/oauth/OAuthCredentialsCache.java
+++ b/clients/java/src/main/java/io/camunda/zeebe/client/impl/oauth/OAuthCredentialsCache.java
@@ -22,6 +22,8 @@ import com.fasterxml.jackson.core.type.TypeReference;
 import com.fasterxml.jackson.databind.ObjectMapper;
 import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
 import io.camunda.zeebe.client.impl.ZeebeClientCredentials;
+import io.camunda.zeebe.client.impl.util.FunctionWithIO;
+import io.camunda.zeebe.client.impl.util.SupplierWithIO;
 import java.io.File;
 import java.io.IOException;
 import java.nio.file.Files;
@@ -74,6 +76,41 @@ public final class OAuthCredentialsCache {
 
   public synchronized Optional<ZeebeClientCredentials> get(final String endpoint) {
     return Optional.ofNullable(audiences.get(endpoint)).map(OAuthCachedCredentials::getCredentials);
+  }
+
+  public synchronized ZeebeClientCredentials computeIfMissingOrInvalid(
+      final String endpoint,
+      final SupplierWithIO<ZeebeClientCredentials> zeebeClientCredentialsConsumer)
+      throws IOException {
+    final Optional<ZeebeClientCredentials> optionalCredentials =
+        readCache()
+            .get(endpoint)
+            .flatMap(
+                zeebeClientCredentials -> {
+                  if (!zeebeClientCredentials.isValid()) {
+                    return Optional.empty();
+                  } else {
+                    return Optional.of(zeebeClientCredentials);
+                  }
+                });
+    if (optionalCredentials.isPresent()) {
+      return optionalCredentials.get();
+    } else {
+      final ZeebeClientCredentials credentials = zeebeClientCredentialsConsumer.get();
+      put(endpoint, credentials).writeCache();
+      return credentials;
+    }
+  }
+
+  public synchronized <T> Optional<T> withCache(
+      final String endpoint, final FunctionWithIO<ZeebeClientCredentials, T> function)
+      throws IOException {
+    final Optional<ZeebeClientCredentials> optionalCredentials = readCache().get(endpoint);
+    if (optionalCredentials.isPresent()) {
+      return Optional.ofNullable(function.apply(optionalCredentials.get()));
+    } else {
+      return Optional.empty();
+    }
   }
 
   public synchronized OAuthCredentialsCache put(

--- a/clients/java/src/main/java/io/camunda/zeebe/client/impl/oauth/OAuthCredentialsProvider.java
+++ b/clients/java/src/main/java/io/camunda/zeebe/client/impl/oauth/OAuthCredentialsProvider.java
@@ -40,9 +40,6 @@ import java.nio.charset.StandardCharsets;
 import java.time.Duration;
 import java.util.HashMap;
 import java.util.Map;
-import java.util.Optional;
-import java.util.concurrent.locks.Lock;
-import java.util.concurrent.locks.ReentrantLock;
 import java.util.stream.Collectors;
 import javax.annotation.concurrent.ThreadSafe;
 import org.slf4j.Logger;
@@ -71,11 +68,6 @@ public final class OAuthCredentialsProvider implements CredentialsProvider {
   private final Duration connectionTimeout;
   private final Duration readTimeout;
 
-  private ZeebeClientCredentials credentials;
-
-  private final Lock fetchCredentialsLock = new ReentrantLock();
-  private final Lock cacheLock = new ReentrantLock();
-
   OAuthCredentialsProvider(final OAuthCredentialsProviderBuilder builder) {
     authorizationServerUrl = builder.getAuthorizationServer();
     endpoint = builder.getAudience();
@@ -88,18 +80,18 @@ public final class OAuthCredentialsProvider implements CredentialsProvider {
   /** Adds an access token to the Authorization header of a gRPC call. */
   @Override
   public void applyCredentials(final Metadata headers) throws IOException {
-    if (credentials == null) {
-      loadCredentials();
-    }
+    final ZeebeClientCredentials zeebeClientCredentials =
+        credentialsCache.computeIfMissingOrInvalid(endpoint, this::fetchCredentials);
 
-    String type = credentials.getTokenType();
+    String type = zeebeClientCredentials.getTokenType();
     if (type == null || type.isEmpty()) {
       throw new IOException(
           String.format("Expected valid token type but was absent or invalid '%s'", type));
     }
 
     type = Character.toUpperCase(type.charAt(0)) + type.substring(1);
-    headers.put(HEADER_AUTH_KEY, String.format("%s %s", type, credentials.getAccessToken()));
+    headers.put(
+        HEADER_AUTH_KEY, String.format("%s %s", type, zeebeClientCredentials.getAccessToken()));
   }
 
   /**
@@ -110,67 +102,19 @@ public final class OAuthCredentialsProvider implements CredentialsProvider {
   public boolean shouldRetryRequest(final Throwable throwable) {
     try {
       return Status.fromThrowable(throwable).getCode() == Code.UNAUTHENTICATED
-          && refreshCredentials();
+          && credentialsCache
+              .withCache(
+                  endpoint,
+                  value -> {
+                    final ZeebeClientCredentials fetchedCredentials = fetchCredentials();
+                    credentialsCache.put(endpoint, fetchedCredentials).writeCache();
+                    return !fetchedCredentials.equals(value) || !value.isValid();
+                  })
+              .orElse(true);
     } catch (final IOException e) {
       LOG.error("Failed while fetching credentials: ", e);
       return false;
     }
-  }
-
-  /** Attempt to load credentials from cache and, if unsuccessful, fetch new credentials. */
-  private void loadCredentials() throws IOException {
-    Optional<ZeebeClientCredentials> cachedCredentials;
-
-    cacheLock.lock();
-    try {
-      try {
-        cachedCredentials = credentialsCache.readCache().get(endpoint);
-      } catch (final IOException e) {
-        LOG.debug("Failed to read credentials cache", e);
-        cachedCredentials = Optional.empty();
-      }
-
-      if (cachedCredentials.isPresent() && cachedCredentials.get().isValid()) {
-        credentials = cachedCredentials.get();
-      } else {
-        refreshCredentials();
-      }
-    } finally {
-      cacheLock.unlock();
-    }
-  }
-
-  /**
-   * Fetch new credentials from authorization server and store them in cache.
-   *
-   * @return true if the fetched credentials are different from the previously stored ones,
-   *     otherwise returns false.
-   */
-  private boolean refreshCredentials() throws IOException {
-    fetchCredentialsLock.lock();
-    final ZeebeClientCredentials fetchedCredentials;
-    try {
-      fetchedCredentials = fetchCredentials();
-    } finally {
-      fetchCredentialsLock.unlock();
-    }
-    final boolean isNewLockCreated = cacheLock.tryLock();
-    try {
-      credentialsCache.put(endpoint, fetchedCredentials).writeCache();
-    } finally {
-      if (isNewLockCreated) {
-        cacheLock.unlock();
-      }
-    }
-
-    if (credentials == null || !credentials.isValid() || !fetchedCredentials.equals(credentials)) {
-      credentials = fetchedCredentials;
-      LOG.debug("Refreshed credentials.");
-
-      return true;
-    }
-
-    return false;
   }
 
   private static String createParams(final OAuthCredentialsProviderBuilder builder) {

--- a/clients/java/src/main/java/io/camunda/zeebe/client/impl/oauth/OAuthCredentialsProvider.java
+++ b/clients/java/src/main/java/io/camunda/zeebe/client/impl/oauth/OAuthCredentialsProvider.java
@@ -41,10 +41,20 @@ import java.time.Duration;
 import java.util.HashMap;
 import java.util.Map;
 import java.util.Optional;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
 import java.util.stream.Collectors;
+import javax.annotation.concurrent.ThreadSafe;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+/**
+ * This class is thread-safe in terms of the next: 1. If you are trying to modify headers of your
+ * request from the several threads there would be sequential calls to the cache 2. If the cache
+ * hasn't a valid token and you are calling from several threads there would be just one call to the
+ * Auth server
+ */
+@ThreadSafe
 public final class OAuthCredentialsProvider implements CredentialsProvider {
   private static final ObjectMapper JSON_MAPPER =
       new ObjectMapper().configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
@@ -62,6 +72,9 @@ public final class OAuthCredentialsProvider implements CredentialsProvider {
   private final Duration readTimeout;
 
   private ZeebeClientCredentials credentials;
+
+  private final Lock fetchCredentialsLock = new ReentrantLock();
+  private final Lock cacheLock = new ReentrantLock();
 
   OAuthCredentialsProvider(final OAuthCredentialsProviderBuilder builder) {
     authorizationServerUrl = builder.getAuthorizationServer();
@@ -108,17 +121,22 @@ public final class OAuthCredentialsProvider implements CredentialsProvider {
   private void loadCredentials() throws IOException {
     Optional<ZeebeClientCredentials> cachedCredentials;
 
+    cacheLock.lock();
     try {
-      cachedCredentials = credentialsCache.readCache().get(endpoint);
-    } catch (final IOException e) {
-      LOG.debug("Failed to read credentials cache", e);
-      cachedCredentials = Optional.empty();
-    }
+      try {
+        cachedCredentials = credentialsCache.readCache().get(endpoint);
+      } catch (final IOException e) {
+        LOG.debug("Failed to read credentials cache", e);
+        cachedCredentials = Optional.empty();
+      }
 
-    if (cachedCredentials.isPresent() && cachedCredentials.get().isValid()) {
-      credentials = cachedCredentials.get();
-    } else {
-      refreshCredentials();
+      if (cachedCredentials.isPresent() && cachedCredentials.get().isValid()) {
+        credentials = cachedCredentials.get();
+      } else {
+        refreshCredentials();
+      }
+    } finally {
+      cacheLock.unlock();
     }
   }
 
@@ -129,8 +147,21 @@ public final class OAuthCredentialsProvider implements CredentialsProvider {
    *     otherwise returns false.
    */
   private boolean refreshCredentials() throws IOException {
-    final ZeebeClientCredentials fetchedCredentials = fetchCredentials();
-    credentialsCache.put(endpoint, fetchedCredentials).writeCache();
+    fetchCredentialsLock.lock();
+    final ZeebeClientCredentials fetchedCredentials;
+    try {
+      fetchedCredentials = fetchCredentials();
+    } finally {
+      fetchCredentialsLock.unlock();
+    }
+    final boolean isNewLockCreated = cacheLock.tryLock();
+    try {
+      credentialsCache.put(endpoint, fetchedCredentials).writeCache();
+    } finally {
+      if (isNewLockCreated) {
+        cacheLock.unlock();
+      }
+    }
 
     if (credentials == null || !credentials.isValid() || !fetchedCredentials.equals(credentials)) {
       credentials = fetchedCredentials;

--- a/clients/java/src/main/java/io/camunda/zeebe/client/impl/util/FunctionWithIO.java
+++ b/clients/java/src/main/java/io/camunda/zeebe/client/impl/util/FunctionWithIO.java
@@ -1,0 +1,23 @@
+/*
+ * Copyright © 2017 camunda services GmbH (info@camunda.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.camunda.zeebe.client.impl.util;
+
+import java.io.IOException;
+
+public interface FunctionWithIO<T, R> {
+
+  R apply(final T value) throws IOException;
+}

--- a/clients/java/src/main/java/io/camunda/zeebe/client/impl/util/SupplierWithIO.java
+++ b/clients/java/src/main/java/io/camunda/zeebe/client/impl/util/SupplierWithIO.java
@@ -1,0 +1,23 @@
+/*
+ * Copyright © 2017 camunda services GmbH (info@camunda.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.camunda.zeebe.client.impl.util;
+
+import java.io.IOException;
+
+@FunctionalInterface
+public interface SupplierWithIO<T> {
+  T get() throws IOException;
+}

--- a/clients/java/src/test/java/io/camunda/zeebe/client/OAuthCredentialsProviderTest.java
+++ b/clients/java/src/test/java/io/camunda/zeebe/client/OAuthCredentialsProviderTest.java
@@ -26,16 +26,21 @@ import static io.camunda.zeebe.client.impl.oauth.OAuthCredentialsProviderBuilder
 import static org.assertj.core.api.Assertions.assertThatThrownBy;
 import static org.assertj.core.api.AssertionsForInterfaceTypes.assertThat;
 import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.spy;
 import static org.mockito.Mockito.times;
 import static org.mockito.Mockito.verify;
 
 import com.github.tomakehurst.wiremock.client.WireMock;
 import com.github.tomakehurst.wiremock.junit.WireMockRule;
+import com.github.tomakehurst.wiremock.matching.RequestPatternBuilder;
+import io.camunda.zeebe.client.api.ZeebeFuture;
 import io.camunda.zeebe.client.api.command.ClientException;
+import io.camunda.zeebe.client.api.response.Topology;
 import io.camunda.zeebe.client.impl.ZeebeClientBuilderImpl;
 import io.camunda.zeebe.client.impl.ZeebeClientCredentials;
 import io.camunda.zeebe.client.impl.ZeebeClientImpl;
 import io.camunda.zeebe.client.impl.oauth.OAuthCredentialsCache;
+import io.camunda.zeebe.client.impl.oauth.OAuthCredentialsProvider;
 import io.camunda.zeebe.client.impl.oauth.OAuthCredentialsProviderBuilder;
 import io.camunda.zeebe.client.impl.util.Environment;
 import io.camunda.zeebe.client.impl.util.EnvironmentRule;
@@ -60,7 +65,11 @@ import java.time.Instant;
 import java.time.ZoneId;
 import java.time.ZonedDateTime;
 import java.time.temporal.ChronoField;
+import java.util.ArrayList;
 import java.util.HashMap;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
 import java.util.function.BiConsumer;
 import java.util.stream.Collectors;
 import org.junit.After;
@@ -278,7 +287,7 @@ public final class OAuthCredentialsProviderTest {
     // then
     assertThat(recordingInterceptor.getCapturedHeaders().get(AUTH_KEY))
         .isEqualTo(TOKEN_TYPE + " " + ACCESS_TOKEN);
-    verify(0, postRequestedFor(WireMock.urlPathEqualTo("/oauth/token")));
+    verify(0, oauthRequestMatcher());
   }
 
   @Test
@@ -300,7 +309,7 @@ public final class OAuthCredentialsProviderTest {
 
     // when
     client.newTopologyRequest().send().join();
-    verify(1, postRequestedFor(WireMock.urlPathEqualTo("/oauth/token")));
+    verify(1, oauthRequestMatcher());
 
     builder.usePlaintext().credentialsProvider(credsBuilder.build());
     client = new ZeebeClientImpl(builder, serverRule.getChannel());
@@ -309,7 +318,7 @@ public final class OAuthCredentialsProviderTest {
     // then
     assertThat(recordingInterceptor.getCapturedHeaders().get(AUTH_KEY))
         .isEqualTo(TOKEN_TYPE + " " + ACCESS_TOKEN);
-    verify(1, postRequestedFor(WireMock.urlPathEqualTo("/oauth/token")));
+    verify(1, oauthRequestMatcher());
     assertCacheContents(cachePath);
   }
 
@@ -349,7 +358,7 @@ public final class OAuthCredentialsProviderTest {
     // then
     assertThat(recordingInterceptor.getCapturedHeaders().get(AUTH_KEY))
         .isEqualTo(TOKEN_TYPE + " " + ACCESS_TOKEN);
-    verify(1, postRequestedFor(WireMock.urlPathEqualTo("/oauth/token")));
+    verify(1, oauthRequestMatcher());
     assertCacheContents(cachePath);
   }
 
@@ -509,6 +518,42 @@ public final class OAuthCredentialsProviderTest {
         .isInstanceOf(IllegalArgumentException.class);
   }
 
+  @Test
+  public void shouldCallOauthServerOnlyOnceInMultithreadMode() throws IOException {
+    // given
+    mockCredentials(ACCESS_TOKEN);
+
+    final OAuthCredentialsProvider credentialsProvider =
+        spy(
+            new OAuthCredentialsProviderBuilder()
+                .clientId(CLIENT_ID)
+                .clientSecret(SECRET)
+                .audience(AUDIENCE)
+                .authorizationServerUrl("http://localhost:" + wireMockRule.port() + "/oauth/token")
+                .credentialsCachePath(tempFolder.newFile().getPath())
+                .build());
+    final ZeebeClientBuilderImpl builder = new ZeebeClientBuilderImpl();
+    builder.usePlaintext().credentialsProvider(credentialsProvider);
+    client = new ZeebeClientImpl(builder, serverRule.getChannel());
+
+    // when
+    final List<ZeebeFuture<Topology>> responses = new ArrayList<>(10);
+    for (int count = 0; count < 10; count++) {
+      responses.add(client.newTopologyRequest().send());
+    }
+    CompletableFuture.allOf(
+            responses.stream()
+                .map(CompletionStage::toCompletableFuture)
+                .collect(Collectors.toList())
+                .toArray(new CompletableFuture[] {}))
+        .join();
+
+    // then
+    assertThat(recordingInterceptor.getCapturedHeaders().get(AUTH_KEY))
+        .isEqualTo(TOKEN_TYPE + " " + ACCESS_TOKEN);
+    verify(1, oauthRequestMatcher());
+  }
+
   /**
    * Mocks an authorization server that returns credentials with the provided access token. Returns
    * the credentials to be return by the server.
@@ -563,6 +608,10 @@ public final class OAuthCredentialsProviderTest {
                             + "\"}")
                     .withFixedDelay(replyDelay)
                     .withStatus(200)));
+  }
+
+  private static RequestPatternBuilder oauthRequestMatcher() {
+    return postRequestedFor(WireMock.urlPathEqualTo("/oauth/token"));
   }
 
   private static String encode(final String param) {

--- a/clients/java/src/test/java/io/camunda/zeebe/client/impl/oauth/OAuthCredentialsCacheTest.java
+++ b/clients/java/src/test/java/io/camunda/zeebe/client/impl/oauth/OAuthCredentialsCacheTest.java
@@ -25,6 +25,13 @@ import java.io.File;
 import java.io.IOException;
 import java.io.InputStream;
 import java.nio.file.Files;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
 import org.junit.Before;
 import org.junit.Rule;
 import org.junit.Test;
@@ -216,5 +223,44 @@ public final class OAuthCredentialsCacheTest {
     assertThat(copy.get(WOMBAT_ENDPOINT)).contains(WOMBAT);
     assertThat(copy.get(AARDVARK_ENDPOINT)).contains(AARDVARK);
     assertThat(copy.size()).isEqualTo(2);
+  }
+
+  @Test
+  public void shouldBeThreadSafe() throws InterruptedException {
+    // given
+    final OAuthCredentialsCache cache = new OAuthCredentialsCache(cacheFile);
+
+    final int threads = 5;
+    final List<Callable<Object>> cacheOperations = new ArrayList<>();
+
+    for (int i = 0; i < 5; i++) {
+      cacheOperations.add(() -> cache.readCache().get(WOMBAT_ENDPOINT));
+      cacheOperations.add(
+          () -> {
+            final OAuthCredentialsCache credentialsCache = cache.put(WOMBAT_ENDPOINT, WOMBAT);
+            credentialsCache.writeCache();
+            return credentialsCache;
+          });
+    }
+
+    final ExecutorService pool = Executors.newFixedThreadPool(threads);
+    try {
+      // when
+      pool.invokeAll(cacheOperations)
+          .forEach(
+              future -> {
+                try {
+                  future.get();
+                } catch (final InterruptedException | ExecutionException e) {
+                  throw new RuntimeException(e);
+                }
+              });
+    } finally {
+      pool.shutdownNow();
+      pool.awaitTermination(60, TimeUnit.SECONDS);
+    }
+
+    // then
+    assertThat(cache.get(WOMBAT_ENDPOINT)).isNotEmpty().contains(WOMBAT);
   }
 }

--- a/clients/java/src/test/java/io/camunda/zeebe/client/impl/oauth/OAuthCredentialsCacheTest.java
+++ b/clients/java/src/test/java/io/camunda/zeebe/client/impl/oauth/OAuthCredentialsCacheTest.java
@@ -234,13 +234,22 @@ public final class OAuthCredentialsCacheTest {
     final List<Callable<Object>> cacheOperations = new ArrayList<>();
 
     for (int i = 0; i < 5; i++) {
-      cacheOperations.add(() -> cache.readCache().get(WOMBAT_ENDPOINT));
       cacheOperations.add(
-          () -> {
-            final OAuthCredentialsCache credentialsCache = cache.put(WOMBAT_ENDPOINT, WOMBAT);
-            credentialsCache.writeCache();
-            return credentialsCache;
-          });
+          () ->
+              cache.computeIfMissingOrInvalid(
+                  WOMBAT_ENDPOINT,
+                  () -> {
+                    cache.put(WOMBAT_ENDPOINT, WOMBAT).writeCache();
+                    return WOMBAT;
+                  }));
+      cacheOperations.add(
+          () ->
+              cache.withCache(
+                  WOMBAT_ENDPOINT,
+                  value -> {
+                    cache.put(WOMBAT_ENDPOINT, WOMBAT).writeCache();
+                    return WOMBAT;
+                  }));
     }
 
     final ExecutorService pool = Executors.newFixedThreadPool(threads);

